core/che-core-api-core/src/main/java/org/eclipse/che/security/PBKDF2PasswordEncryptor.java

```diff
@@ -0,0 +1,90 @@
+ *******************************************************************************/
+package org.eclipse.che.security;
+
+import com.google.common.hash.HashCode;
+
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.primitives.Ints.tryParse;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Encrypts password using <a href="https://en.wikipedia.org/wiki/PBKDF2">Password-based-Key-Derivative-Function</a>
+ * with <b>SHA512</b> as pseudorandom function.
+ * See <a href="https://www.ietf.org/rfc/rfc2898.txt">rfc2898</a>.
+ *
+ * @author Yevhenii Voevodin
+ */
+public class PBKDF2PasswordEncryptor implements PasswordEncryptor {
+
+    private static final String  PWD_FMT   = "%s:%s:%d";
+    private static final Pattern PWD_REGEX = Pattern.compile("(?<pwdHash>\\w+):(?<saltHash>\\w+):(?<iterations>\\w+)");
+
+    private static final String       SECRET_KEY_FACTORY_NAME = "PBKDF2WithHmacSHA512";
+    private static final SecureRandom SECURE_RANDOM           = new SecureRandom();
+    /**
+     * Minimum number of iterations required is 1_000(rfc2898),
+     * pick greater as potentially safer in the case of brute-force attacks .
+     */
+    private static final int          ITERATIONS_COUNT        = 10_000;
+    /** 64bit salt length based on the rfc2898 spec . */
+    private static final int          SALT_LENGTH             = 64 / 8;
+
+    @Override
+    public String encrypt(String password) {
+        requireNonNull(password, "Required non-null password");
+        final byte[] salt = new byte[SALT_LENGTH];
+        SECURE_RANDOM.nextBytes(salt);
+        final HashCode hash = computeHash(password.toCharArray(), salt, ITERATIONS_COUNT);
+        final HashCode saltHash = HashCode.fromBytes(salt);
+        return format(PWD_FMT, hash, saltHash, ITERATIONS_COUNT);
+    }
+
+    @Override
+    public boolean test(String password, String encryptedPassword) {
+        requireNonNull(password, "Required non-null password");
+        requireNonNull(password, "Required non-null encrypted password");
+        final Matcher matcher = PWD_REGEX.matcher(encryptedPassword);
+        if (!matcher.matches()) {
+            return false;
+        }
+        // retrieve salt, password hash and iterations count from hash
+        final Integer iterations = tryParse(firstNonNull(matcher.group("iterations"), ""));
+        final String salt = matcher.group("saltHash");
+        final String pwdHash = matcher.group("pwdHash");
+        if (salt == null || iterations == null || iterations <= 0 || pwdHash == null) {
```

Comment:
Encrypted password matches regexp and regexp require 1+ symbols in each group. It looks like NULL checks are not needed.